### PR TITLE
PX4 bus threading

### PR DESCRIPTION
**Pull request by tridge**
This adds I2C and SPI thread-per-bus support on PX4.
This is just the infrastructure, there will be a separate PR to enable the use of the new implementation for IMU, mag, baro etc.
Note that this does fix the long-standing issue with the I2C interface for the LightWare lidar

### CODE REVIEW

**Comment by tridge** [comment]
@lucasdemarchi, note the method I used for MPU6000 to cope with not having register_periodic_callback() on AuxiliaryBus yet.

File: libraries/AP_HAL_PX4/I2CDevice.cpp [hunk 1]
```diff
@@ -17,16 +17,54 @@
...
+
+/*
+  map ArduPilot bus numbers to PX4 bus numbers
+ */    
+uint8_t PX4_I2C::map_bus_number(uint8_t bus) const
+{
+    switch (bus) {
+    case 0:
+        // map to internal bus
+#ifdef PX4_I2C_BUS_ONBOARD
```
Comment (lucasdemarchi) [inline libraries/AP_HAL_PX4/I2CDevice.cpp]: Also, if it's not defined, shouldn't you return an invalid number to mean "bus not available"?  Looking at the source code PX4-V4 doesn't have an onboard bus, only an external one... AFAICS we would return a wrong number here.

File: libraries/AP_HAL_PX4/I2CDevice.cpp [hunk 2]
```diff
@@ -67,6 +107,122 @@ bool I2CDevice::read_registers_multiple(uint8_t first_reg, uint8_t *recv,
...
+    struct sched_param param;
+    
+    pthread_attr_init(&thread_attr);
+    pthread_attr_setstacksize(&thread_attr, 1024);
+    
+    param.sched_priority = APM_I2C_PRIORITY;
+    (void)pthread_attr_setschedparam(&thread_attr, &param);
+    pthread_attr_setschedpolicy(&thread_attr, SCHED_FIFO);
+    
+    pthread_create(&binfo.thread_ctx, &thread_attr, &I2CDevice::i2c_thread, &binfo);
```
Comment (lucasdemarchi) [inline libraries/AP_HAL_PX4/I2CDevice.cpp]: Humn... maybe having a Thread implementation just like in the Linux layer would be good to reduce the boilerplate code to create/start/stop a thread.

File: libraries/AP_HAL_PX4/Scheduler.h [hunk 3]
```diff
@@ -13,7 +13,7 @@
 #define APM_MAIN_PRIORITY_BOOST 241
 #define APM_MAIN_PRIORITY       180
 #define APM_TIMER_PRIORITY      181
-#define APM_SPI_PRIORITY        182
+#define APM_SPI_PRIORITY        242
```
Comment (lucasdemarchi) [inline libraries/AP_HAL_PX4/Scheduler.h]: In other words, you always guarantee main thread has CPU available for completing its job, and have it to wait and boost the priority of whatever it's waiting for.

**Comment by lucasdemarchi** [comment]
> @lucasdemarchi, note the method I used for MPU6000 to cope with not having register_periodic_callback() on AuxiliaryBus yet.

I need to test this... I can't see how it would work since `_dev` in MPU6000 is a SPIDevice, which does support `register_periodic_callback()`. Didn't you mistakenly add it in MPU6000 rather than in HMC5843?

File: libraries/AP_HAL_PX4/I2CDevice.cpp (hunk 2 above)
Comment (tridge) [inline libraries/AP_HAL_PX4/I2CDevice.cpp]: yep, done now, thanks

**Comment by tridge** [comment]
> Didn't you mistakenly added it in MPU6000 rather than in HMC5843?

I don't think so. Inside hmc driver we have _bus of type AuxiliaryBus* in the AP_HMC5843_BusDriver_Auxiliary class. We need to ask that bus to add a periodic callback. So we have to extend AuxiliaryBus to understand that.

**Comment by OXINARF** [comment]
@tridge This was merged together with #5109 right?

**Comment by tridge** [comment]
yep!

**Pull request closed without merging**